### DISCUSSION TITLE
Language Client in Static Page

### DISCUSSION THREAD

**Discussion by leezetech** [TypeFox/monaco-languageclient#796] [Q&A]
Hi, I'm very much a newbie to this and would very much appreciate some help.  I am currently running the vanilla monaco editor locally and statically in a webview2 control in .NET by simply navigating to the index.html file on the local drive that creates an instance of the editor. The interface is very simple and allows me to to run javascript to get data in/out of the editor between the browser and .NET.
I have been attempting to think about how I might use the language client in this scenario but it appears that the only way to do this would be to run the client wrapper in a node application on a server? Is this true or is there a way to run the language client in a static page?
Many thanks

**Comment by kaisalmen** [comment]
Hi @leezetech yes you can run the client in a static pages. JavaScript must be enabled/available, of course. The language server can either be run inside a web worker (not always possible) or as an externally process the client connects to via web socket. Our examples that don't require a backend are available via GitHub Pages here:
https://typefox.github.io/monaco-languageclient/index.html

An overview of all example you find here:
https://github.com/TypeFox/monaco-languageclient?tab=readme-ov-file#main-examples

**Reply by leezetech** [reply]
Thankyou very much for replying. I downloaded the repository and built it so I ended up with a 'dist' folder in the example folders. I'm assuming that the GitHub Pages examples referenced above are a subset of the main examples that can be run without a back end?
I've then looked at the 'browser' example (from my now local repository which I think is known as the 'Language Client Pure Browser Example' when run from the GitHubPages) and attempted to replicate the GitHub Pages version on my local machine.

I'm running into a few issues which I'm not sure how to resolve:

1. I was expecting to see a browser.html file in the 'Monaco\monaco-languageclient-main\packages\examples\dist' folder that referenced the built main.js file, as the one in the examples folder still references the typescript file 'main.ts' and points to the source files:

import { runBrowserEditor } from "./src/browser/main.ts"

2. When I try to create an equivalent browser.html file in the 'dist' folder and point the script as follows:

import { runBrowserEditor } from "./browser/main.js"

I can get it to run locally but there are the following errors:

![Errors](https://github.com/user-attachments/assets/deee4a7e-795c-4507-97a7-4b90ca5ea20b)

I suspect I don't really understand what I'm doing, as I don't think the build command has packaged the dependencies in such a way that I can run the example locally on my hard disk. Apologies for my ignorance  - I would greatly appreciate some advice on how to get this working,
Many Thanks

**Reply by kaisalmen** [reply]
The things available via GitHub Pages are created by the vite production build:
https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/package.json#L137
Artifacts end up in packages/examples/production and not in dist

The vite dev server (see https://github.com/TypeFox/monaco-languageclient/tree/main#getting-started) is also statically serving the files in the end. If you cloned the repo you can just do waht is described in getting started and change whatever you want.

**Reply by leezetech** [reply]
Thankyou for your response, I don't have anything in the production folder so I obviously haven't done this correctly. I'll have another try, many thanks

**Reply by kaisalmen** [reply]
@leezetech you have to run `npm run production:build`

**Reply by leezetech** [reply]
Thank you very much, I was using an older version which didn't have the production commands. I now have a production folder with an assets directory. However I've noticed that it appears that all the assets relating to the 'browser' example are there in the assets folder but everything still seems to be linked to the files on the internet as opposed to the local assets, so it won't run without an internet connection. Is there a way of pointing all of the links to the asset folder instead?
Also, there are approx 1500 files in the assets folder - is there a way to bundle these files into a smaller number of working files and update the internal dependency structure?
Many thanks

**Reply by kaisalmen** [reply]
@leezetech there is a local configuration for the production preview and the one that the production build is using. The local preview is build and started `production:preview:build`and `production:preview`, see here:
https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/package.json#L138-L139

> Also, there are approx 1500 files in the assets folder - is there a way to bundle these files into a smaller number of working files and update the internal dependency structure?

You have to play around with vite/rollup setttigs

**Reply by leezetech** [reply]
Thanks very much for your quick replies, I will look into it.

**Reply by leezetech** [reply]
Hi thank you for your previous help, I've managed to get the browser example running locally. Ultimately I'm attempting to get a Python language client running locally so could I just ask why, at the moment, the Python client example is not considered to be able to run statically and locally?
Many thanks

**Reply by kaisalmen** [reply]
@leezetech yes, you can it run locally. The example on the GitHub Pages are limited to those that don't require a backend. The vite dev server "serves" all client statically. You can change the production config in your fork to include other example. You just need to specify a reachable language server address the client can use.
